src/emulator/storage/apis/firebase.ts

```diff
@@ -323,186 +324,147 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
+      res.header("x-goog-upload-chunk-granularity", "10000");
+      res.header("x-goog-upload-control-url", "");
+      res.header("x-goog-upload-status", "active");
+      const emulatorInfo = EmulatorRegistry.getInfo(Emulators.STORAGE);
+      res.header(
+        "x-goog-upload-url",
+        `http://${req.hostname}:${emulatorInfo?.port}/v0/b/${bucketId}/o?name=${objectId}&upload_id=${upload.id}&upload_protocol=resumable`
+      );
+      res.header("x-gupload-uploadid", upload.id);
 
-        res.status(200).send();
-        return;
-      }
+      return res.sendStatus(200);
+    }
 
-      if (!req.query.upload_id) {
-        res.sendStatus(400);
-        return;
-      }
+    if (!req.query.upload_id) {
+      return res.sendStatus(400);
+    }
 
-      const uploadId = req.query.upload_id.toString();
-      if (uploadCommand === "query") {
-        const upload = storageLayer.queryUpload(uploadId);
-        if (!upload) {
-          res.sendStatus(400);
-          return;
+    const uploadId = req.query.upload_id.toString();
+    if (uploadCommand === "query") {
+      let upload: Upload;
+      try {
+        upload = uploadService.getResumableUpload(uploadId);
+      } catch (err) {
+        if (err instanceof NotFoundError) {
+          return res.sendStatus(404);
         }
-
-        res.header("X-Goog-Upload-Size-Received", upload.currentBytesUploaded.toString());
-        res.sendStatus(200);
-        return;
+        throw err;
       }
+      res.header("X-Goog-Upload-Size-Received", upload.size.toString());
+      return res.sendStatus(200);
+    }
 
-      if (uploadCommand === "cancel") {
-        const upload = storageLayer.queryUpload(uploadId);
-        if (upload) {
-          const cancelled = storageLayer.cancelUpload(upload);
-          res.sendStatus(cancelled ? 200 : 400);
-        } else {
-          res.sendStatus(404);
+    if (uploadCommand === "cancel") {
+      try {
+        uploadService.cancelResumableUpload(uploadId);
+      } catch (err) {
+        if (err instanceof NotFoundError) {
+          return res.sendStatus(404);
+        } else if (err instanceof NotCancellableError) {
+          return res.sendStatus(400);
         }
-        return;
+        throw err;
       }
+      return res.sendStatus(200);
+    }
 
-      let upload;
-      if (uploadCommand.includes("upload")) {
-        upload = storageLayer.uploadBytes(uploadId, await reqBodyToBuffer(req));
-
-        if (!upload) {
-          res.sendStatus(400);
-          return;
+    if (uploadCommand.includes("upload")) {
+      let upload: Upload;
+      try {
+        upload = uploadService.continueResumableUpload(uploadId, await reqBodyToBuffer(req));
+      } catch (err) {
+        if (err instanceof NotFoundError) {
+          return res.sendStatus(404);
+        } else if (err instanceof UploadNotActiveError) {
+          return res.sendStatus(400);
         }
-
+        throw err;
+      }
+      if (!uploadCommand.includes("finalize")) {
```

Comment:
A little odd that you check the same thing here and in line 433. I'd just do the fallthrough?

Reply:
The check is to prevent us falling through to the unsupported method case and returning a 400. It is a little weird though, if you don't mind, I'd like to keep this for now and refactor the structure of this method in a follow up PR. It's already a little hard to tell the deltas since so much is changing, I'd like to keep this scoped to just behavioral changes for now.